programs/psa/aead_cipher_psa.c

```diff
@@ -0,0 +1,352 @@
+            "MBEDTLS_CHACHAPOLY_C not defined, and/or "
+            "MBEDTLS_PSA_CRYPTO_KEY_ID_ENCODES_OWNER defined\r\n" );
+    return( 0 );
+}
+#else
+
+#include <string.h>
+
+#include "mbedtls/cipher.h"
+#include "psa/crypto.h"
+
+/*
+ * Common data and helper functions
+ */
+const char usage[] = "Usage: aead_cipher_psa [gcm128|gcm256|gcm128_8|chachapoly]";
+
+const unsigned char iv1[12] = { 0x00 };
+const unsigned char ad1[] = { 0x01, 0x02 };
+const unsigned char pa1[] = { 0x03, 0x04 };
+const unsigned char pb1[] = { 0x05, 0x06, 0x07 };
+
+const unsigned char iv2[12] = { 0x10 };
+const unsigned char ad2[] = { 0x11, 0x12 };
+const unsigned char pa2[] = { 0x13, 0x14 };
+const unsigned char pb2[] = { 0x15, 0x16, 0x17 };
+
+const unsigned char key_bytes[32] = { 0x2a };
+
+void print_out( const char *title, unsigned char *out, size_t len )
+{
+    printf( "%s:", title );
+    for( size_t i = 0; i < len; i++ )
+        printf( " %02x", out[i] );
+    printf( "\n" );
+}
+
+/*
+ * Functions using the Cipher API
+ */
+#define CHK( code )         \
+    do {                    \
+        ret = code;         \
+        if( ret != 0 ) {    \
+            printf( "%03d: ret = -0x%04x\n", __LINE__, (unsigned) -ret ); \
+            goto exit;      \
+        }                   \
+    } while( 0 )
+
+
+static int cipher_prepare( const char *info,
+                           mbedtls_cipher_context_t *ctx,
+                           size_t *tag_len )
+{
+    int ret;
+
+    mbedtls_cipher_type_t type;
+    if( strcmp( info, "gcm128" ) == 0 ) {
+        type = MBEDTLS_CIPHER_AES_128_GCM;
+        *tag_len = 16;
+    } else if( strcmp( info, "gcm256" ) == 0 ) {
+        type = MBEDTLS_CIPHER_AES_256_GCM;
+        *tag_len = 16;
+    } else if( strcmp( info, "gcm128_8" ) == 0 ) {
+        type = MBEDTLS_CIPHER_AES_128_GCM;
+        *tag_len = 8;
+    } else if( strcmp( info, "chachapoly" ) == 0 ) {
+        type = MBEDTLS_CIPHER_CHACHA20_POLY1305;
+        *tag_len = 16;
+    } else {
+        puts( usage );
+        return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+    }
+
+    CHK( mbedtls_cipher_setup( ctx,
+                               mbedtls_cipher_info_from_type( type ) ) );
+
+    size_t key_len = mbedtls_cipher_get_key_bitlen( ctx );
+    CHK( mbedtls_cipher_setkey( ctx, key_bytes, key_len, MBEDTLS_ENCRYPT ) );
+
+exit:
+    return( ret );
+}
+
+static void cipher_info( const mbedtls_cipher_context_t *ctx, size_t tag_len )
+{
+    // no convenient way to get the cipher type (for example, AES)
+    const char *ciph = "???";
+    int key_bits = mbedtls_cipher_get_key_bitlen( ctx );
+    mbedtls_cipher_mode_t mode = mbedtls_cipher_get_cipher_mode( ctx );
+
+    const char *mode_str = mode == MBEDTLS_MODE_GCM ? "GCM"
+                         : mode == MBEDTLS_MODE_CHACHAPOLY ? "ChachaPoly"
+                         : "???";
+
+    printf( "cipher: %s, %d, %s, %zu\n", ciph, key_bits, mode_str, tag_len );
```

Comment:
CI says we're still testing with a version of mingw that doesn't support `%zu`. (And grep says we do use it but only in `memory_buffer_alloc`). We use `MBEDTLS_PRINTF_SIZET` instead which has a workaround for old mingw.

Also this should be `mbedtls_printf`, not `printf`.

Reply:
Regarding `zu`, I'm inclined to just cast to `unsigned` there's really no risk of overflow here, and to a reader who it's already familiar with it `MBEDTLS_PRINTF_SIZET` may be a distraction. (We are casting to unsigned in other programs, and so far the only uses of `MBEDTLS_PRINTF_SIZET` are in the library.)

Reply:
> I had a look at the other programs in programs/psa and they're using printf

That dates back from https://github.com/ARMmbed/mbed-crypto/pull/74, which was motivated by the way we were trying to do crypto/tls separation at the time. I'm not sure that's the right decision today.

I do think it would be nicer to stick to standard functions in example code. On the other hand, this limits the code's portability. So the question is, to we care about being able to run sample program as-is on ultra-constrained or exotic platforms that lack e.g. `printf` or `malloc` or `exit`?

Reply:
> I do think it would be nicer to stick to standard functions in example code.

I agree. Although it reduces the sample programs portability, I think it is more useful like that for users. As a user, reading the sample code I would definitely wonder what these `mbedtls_` prefixed functions do in the background. I wouldn't be certain if some library functions depend on that behaviour and if I am allowed to use standard functions.

Users of that ultra-constrained or exotic platforms need to hook up their implementations in the platform abstraction layer anyway in the first place and that couple of find and replace would be a negligible effort.

Reply:
Keeping in mind that “exotic” platforms include not-that-ancient Windows, e.g. this discussion started with an error from using standard functionality on MinGW.

Reply:
Tbh, I never understood why `mbedtls_printf()` and all were introduced in the example programs in the first place.

Most of the programs are mostly educational: the goal is mainly to read them, run them on you development platform if it helps understanding how they work (perhaps with some editing etc.), and perhaps copy-paste bits and pieces to you own program. I can't see why you'll need or want to run them as-is on an embedded device.

Some of the the programs are not just educational, but could also be used as actual utilities, for example the x509 ones. Even then, I'm not convinced running them as-is on embedded devices is a reasonable goal: these programs take command-line options, which hardly make sense in an embedded context, so at least some adaptation will be needed anyway.

Some other programs are for tests (`ssl_xxx2.c`) - this is perhaps the thing you're most likely to want to run on target, but you also have the issue of command-line options, currently the program depend on the sockets API, which I'd say is much more of a constraint the things from libc...

Anyway, at some point we might want to rationalize that, but I hope we can get this PR merged before that :)

> For example, here, how about using PRIuPTR and defining it if it's unset and the implementation is MinGW?

I think that's a reasonable option, but in this case I think using `"%u", (unsigned) size` works just as well, and it's also what we already do in several other places.